File: Visual Basic/Items/MSCE.ElementWin/ElementToolWin.vb
'*--------------------------------------------------------------------------------------+
'   $safeitemname$.vb
'
'+--------------------------------------------------------------------------------------*/

#Region "Bentley Namespaces"
Imports BDPN = Bentley.DgnPlatformNET
Imports BG = Bentley.GeometryNET
Imports BM = Bentley.MstnPlatformNET
Imports BMW = Bentley.MstnPlatformNET.WinForms
#End Region

Class $safeitemname$
		Inherits BDPN.DgnElementSetTool

    'Use $safeitemname$MSForm.Property, $safeitemname$MSForm.Method(), or $safeitemname$MSForm.Field
    'to access public properties, methods, Or fields in $safeitemname$Form.cs.

#Region "ToolSettings"
    Private Shared Property $safeitemname$Command As $safeitemname$
    Private Shared Property $safeitemname$MSForm As $safeitemname$Form
    Private Shared Property $safeitemname$Adapter As BMW.Adapter

    Class MyToolSettings
        Inherits BMW.Adapter
        Private Property m_command As $safeitemname$

        Public Sub New(command As $safeitemname$)
            m_command = command
            $safeitemname$MSForm = New $safeitemname$Form()
        End Sub
    End Class
#End Region

    Public Sub New(ToolId As Integer, toolName As Integer)
        MyBase.New()
        $safeitemname$Adapter = New MyToolSettings(Me)
    End Sub

#Region "DgnElementSetTool Members"
    Protected Overrides Sub OnRestartTool()
        InstallNewInstance()
    End Sub

    Protected Overrides Sub OnCleanup()
        If $safeitemname$Command IsNot Nothing Then
            $safeitemname$MSForm.DetachFromMicroStation()
            $safeitemname$MSForm.Dispose()
            $safeitemname$MSForm = Nothing
            $safeitemname$Command = Nothing
        End If
        MyBase.OnCleanup()
    End Sub

    Public Overrides Function OnElementModify(element As BDPN.Elements.Element) As BDPN.StatusInt

        Return BDPN.StatusInt.Success
    End Function

    Protected Overrides Function OnResetButton(ev As BDPN.DgnButtonEvent) As Boolean
        ExitTool()
        Return True
    End Function

    Protected Overrides Sub ExitTool()

        MyBase.ExitTool()
    End Sub

    Protected Overrides Sub OnDynamicFrame(ByVal ev As BDPN.DgnButtonEvent)
    End Sub

    Protected Overrides Function OnInstall() As Boolean

        Return True
    End Function

    Protected Overrides Sub OnPostInstall()
        MyBase.OnPostInstall()
    End Sub

    Public Sub InstallNewInstance(ByVal Optional unparsed As String = Nothing)
        If $safeitemname$Command Is Nothing Then
            $safeitemname$Command = New $safeitemname$(0, 0)
            $safeitemname$Command.InstallTool()
            $safeitemname$MSForm.Text = "$safeitemname$ Tool Settings"
            $safeitemname$MSForm.AttachToToolSettings(Program.Addin)
        Else
            $safeitemname$MSForm.Focus()
        End If
    End Sub
#End Region
End Class

File: CSharp/Items/MSCE.ElementWin/ElementToolWin.cs
/*--------------------------------------------------------------------------------------+
|   $safeitemname$.cs
|
+--------------------------------------------------------------------------------------*/

#region Bentley Namespaces
using BDPN = Bentley.DgnPlatformNET;
using BG = Bentley.GeometryNET;
using BM = Bentley.MstnPlatformNET;
using BMW = Bentley.MstnPlatformNET.WinForms;
#endregion

namespace $rootnamespace$
{
    class $safeitemname$ : BDPN.DgnElementSetTool
    {
    //Use $safeitemname$MSForm.Property, $safeitemname$MSForm.Method(), or $safeitemname$MSForm.Field
    //to access public properties, methods, or fields in $safeitemname$Form.cs.

     #region ToolSettings
        private static $safeitemname$ $safeitemname$Command { get; set; }
private static $safeitemname$Form $safeitemname$MSForm { get; set; }
private static BMW.Adapter $safeitemname$Adapter { get; set; }

class MyToolSettings : BMW.Adapter
{
    private $safeitemname$ m_command;

    public MyToolSettings($safeitemname$ command)
    {
        m_command = command;
        $safeitemname$MSForm = new $safeitemname$Form();
    }
};
#endregion

public $safeitemname$(int toolID, int toolName) : base()
        {
            $safeitemname$Adapter = new MyToolSettings(this);
}

#region DgnElementSetTool Members
protected override void OnRestartTool()
{
    InstallNewInstance();
}

protected override void OnCleanup()
{
    if (null != $safeitemname$Command)
    {
        $safeitemname$MSForm.DetachFromMicroStation();
        $safeitemname$MSForm.Dispose();
        $safeitemname$MSForm = null;
        $safeitemname$Command = null;
    }
    base.OnCleanup();
}

public override BDPN.StatusInt OnElementModify(BDPN.Elements.Element element)
{

    return BDPN.StatusInt.Success;
}

protected override bool OnResetButton(BDPN.DgnButtonEvent ev)
{
    ExitTool();
    return true;
}

protected override void ExitTool()
{

    base.ExitTool();
}

protected override void OnDynamicFrame(BDPN.DgnButtonEvent ev)
{

}

protected override bool OnInstall()
{

    return true;
}

protected override void OnPostInstall()
{
    base.OnPostInstall();
}

public void InstallNewInstance(string unparsed = "")
{
    if (null == $safeitemname$Command)
    {
        $safeitemname$Command = new $safeitemname$(0, 0);
        $safeitemname$Command.InstallTool();
        $safeitemname$MSForm.AttachToToolSettings(Program.Addin);
        $safeitemname$MSForm.Text = "$safeitemname$ Tool Settings";
    }
    else
        $safeitemname$MSForm.Focus();
}
        #endregion
    }
}